File: demo/twosum.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""Demo: solving the Two Sum problem.

Given a list of integers and a target value, find the two indices whose
elements add up to the target.  Three approaches are shown:

  1. Brute-force  O(n²)
  2. Hash-map      O(n)   ← the classic interview answer
  3. Sorted + two-pointer  O(n log n)  (returns values, not original indices)

Run:
  python twosum.py
  # or, if uv is available:
  uv run twosum.py
"""

from __future__ import annotations


# ---------------------------------------------------------------------------
# Approach 1 – Brute force  O(n²)
# ---------------------------------------------------------------------------

def two_sum_brute(nums: list[int], target: int) -> tuple[int, int] | None:
    """Return (i, j) with i < j such that nums[i] + nums[j] == target."""
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                return (i, j)
    return None


# ---------------------------------------------------------------------------
# Approach 2 – Hash-map  O(n)
# ---------------------------------------------------------------------------

def two_sum_hashmap(nums: list[int], target: int) -> tuple[int, int] | None:
    """Single-pass hash-map lookup.  Returns (i, j) with i < j."""
    seen: dict[int, int] = {}          # value -> index
    for j, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return (seen[complement], j)
        seen[num] = j
    return None


# ---------------------------------------------------------------------------
# Approach 3 – Sorted two-pointer  O(n log n)
#              (works on values; rebuilds original indices at the end)
# ---------------------------------------------------------------------------

def two_sum_two_pointer(nums: list[int], target: int) -> tuple[int, int] | None:
    """Sort a copy, use two pointers, then map back to original indices."""
    indexed = sorted(enumerate(nums), key=lambda x: x[1])  # (orig_idx, val)
    lo, hi = 0, len(indexed) - 1
    while lo < hi:
        s = indexed[lo][1] + indexed[hi][1]
        if s == target:
            i, j = indexed[lo][0], indexed[hi][0]
            return (min(i, j), max(i, j))
        elif s < target:
            lo += 1
        else:
            hi -= 1
    return None


# ---------------------------------------------------------------------------
# Demo runner
# ---------------------------------------------------------------------------

TEST_CASES: list[tuple[list[int], int]] = [
    ([2, 7, 11, 15], 9),          # classic LeetCode example  → (0, 1)
    ([3, 2, 4], 6),               # non-adjacent pair         → (1, 2)
    ([3, 3], 6),                  # duplicate values          → (0, 1)
    ([1, 5, 3, 7, 2, 9], 10),     # longer list               → (1, 3)  5+7
    ([1, 2, 3], 100),             # no solution               → None
]

SOLVERS = [
    ("brute-force  O(n²)",       two_sum_brute),
    ("hash-map     O(n) ",       two_sum_hashmap),
    ("two-pointer  O(n log n)",  two_sum_two_pointer),
]


def main() -> None:
    header = f"{'nums':<30}  {'target':>6}  {'brute':>10}  {'hashmap':>10}  {'2-ptr':>10}"
    print(header)
    print("-" * len(header))

    for nums, target in TEST_CASES:
        results = [solver(list(nums), target) for _, solver in SOLVERS]
        # Verify each result is correct (values sum to target, or both None)
        for (name, _), r in zip(SOLVERS, results):
            if r is None:
                assert all(res is None for res in results), (
                    f"{name}: returned None but another solver found a result"
                )
            else:
                i, j = r
                assert nums[i] + nums[j] == target, (
                    f"{name}: nums[{i}]+nums[{j}]={nums[i]+nums[j]} != {target}"
                )
        cols = "  ".join(f"{str(r):>10}" for r in results)
        print(f"{str(nums):<30}  {target:>6}  {cols}")

    print("\nAll solvers agree on every test case. ✓")

    # -----------------------------------------------------------------
    # Walk through the hash-map approach step-by-step for one example
    # -----------------------------------------------------------------
    print("\n── Step-by-step trace (hash-map, nums=[2,7,11,15], target=9) ──")
    nums_trace = [2, 7, 11, 15]
    target_trace = 9
    seen: dict[int, int] = {}
    for j, num in enumerate(nums_trace):
        complement = target_trace - num
        print(f"  j={j}  num={num}  complement={complement}  seen={seen}")
        if complement in seen:
            i = seen[complement]
            print(f"  → found! indices ({i}, {j})  "
                  f"nums[{i}]+nums[{j}] = {nums_trace[i]}+{nums_trace[j]} = {target_trace}")
            break
        seen[num] = j


if __name__ == "__main__":
    main()

File: typescript_demo/twosum.ts
/**
 * Demo: solving the Two Sum problem.
 *
 * Given a list of integers and a target value, find the two indices whose
 * elements add up to the target. Three approaches are shown:
 *
 *   1. Brute-force            O(n²)
 *   2. Hash-map               O(n)   ← the classic interview answer
 *   3. Sorted + two-pointer   O(n log n)  (maps back to original indices)
 *
 * Run:
 *   npx tsx twosum.ts
 */
import { fileURLToPath } from "node:url";

type Pair = [number, number];

// ---------------------------------------------------------------------------
// Approach 1 – Brute force  O(n²)
// ---------------------------------------------------------------------------
function twoSumBrute(nums: number[], target: number): Pair | null {
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] === target) {
        return [i, j];
      }
    }
  }
  return null;
}

// ---------------------------------------------------------------------------
// Approach 2 – Hash-map  O(n)
// ---------------------------------------------------------------------------
function twoSumHashmap(nums: number[], target: number): Pair | null {
  const seen = new Map<number, number>(); // value -> index
  for (let j = 0; j < nums.length; j++) {
    const complement = target - nums[j];
    if (seen.has(complement)) {
      return [seen.get(complement)!, j];
    }
    seen.set(nums[j], j);
  }
  return null;
}

// ---------------------------------------------------------------------------
// Approach 3 – Sorted two-pointer  O(n log n)
//              (works on values; rebuilds original indices at the end)
// ---------------------------------------------------------------------------
function twoSumTwoPointer(nums: number[], target: number): Pair | null {
  const indexed = nums
    .map((val, idx): Pair => [idx, val])
    .sort((a, b) => a[1] - b[1]); // [origIdx, val]
  let lo = 0;
  let hi = indexed.length - 1;
  while (lo < hi) {
    const sum = indexed[lo][1] + indexed[hi][1];
    if (sum === target) {
      const [i, j] = [indexed[lo][0], indexed[hi][0]];
      return [Math.min(i, j), Math.max(i, j)];
    } else if (sum < target) {
      lo += 1;
    } else {
      hi -= 1;
    }
  }
  return null;
}

// ---------------------------------------------------------------------------
// Demo runner
// ---------------------------------------------------------------------------
const TEST_CASES: [number[], number][] = [
  [[2, 7, 11, 15], 9], // classic LeetCode example  -> (0, 1)
  [[3, 2, 4], 6], //        non-adjacent pair  -> (1, 2)
  [[3, 3], 6], //              duplicate values  -> (0, 1)
  [[1, 5, 3, 7, 2, 9], 10], //       longer list  -> (1, 3)  5+7
  [[1, 2, 3], 100], //               no solution  -> null
];

const SOLVERS: [string, (nums: number[], target: number) => Pair | null][] = [
  ["brute-force  O(n²)", twoSumBrute],
  ["hash-map     O(n) ", twoSumHashmap],
  ["two-pointer  O(n log n)", twoSumTwoPointer],
];

function pad(s: string, width: number, right = false): string {
  return right ? s.padStart(width) : s.padEnd(width);
}

function show(pair: Pair | null): string {
  return pair === null ? "null" : `(${pair[0]}, ${pair[1]})`;
}

function main(): void {
  const header =
    `${pad("nums", 30)}  ${pad("target", 6, true)}  ` +
    `${pad("brute", 10, true)}  ${pad("hashmap", 10, true)}  ${pad("2-ptr", 10, true)}`;
  console.log(header);
  console.log("-".repeat(header.length));

  for (const [nums, target] of TEST_CASES) {
    const results = SOLVERS.map(([, solver]) => solver([...nums], target));
    // Verify each result is correct (values sum to target, or all are null).
    results.forEach((r, idx) => {
      const name = SOLVERS[idx][0];
      if (r === null) {
        if (!results.every((res) => res === null)) {
          throw new Error(`${name}: returned null but another solver found a result`);
        }
      } else {
        const [i, j] = r;
        if (nums[i] + nums[j] !== target) {
          throw new Error(`${name}: nums[${i}]+nums[${j}]=${nums[i] + nums[j]} != ${target}`);
        }
      }
    });
    const cols = results.map((r) => pad(show(r), 10, true)).join("  ");
    console.log(`${pad(`[${nums.join(", ")}]`, 30)}  ${pad(String(target), 6, true)}  ${cols}`);
  }

  console.log("\nAll solvers agree on every test case. ✓");

  // -----------------------------------------------------------------
  // Walk through the hash-map approach step-by-step for one example
  // -----------------------------------------------------------------
  console.log("\n── Step-by-step trace (hash-map, nums=[2,7,11,15], target=9) ──");
  const numsTrace = [2, 7, 11, 15];
  const targetTrace = 9;
  const seen = new Map<number, number>();
  for (let j = 0; j < numsTrace.length; j++) {
    const num = numsTrace[j];
    const complement = targetTrace - num;
    console.log(`  j=${j}  num=${num}  complement=${complement}  seen=${JSON.stringify([...seen])}`);
    if (seen.has(complement)) {
      const i = seen.get(complement)!;
      console.log(
        `  → found! indices (${i}, ${j})  ` +
          `nums[${i}]+nums[${j}] = ${numsTrace[i]}+${numsTrace[j]} = ${targetTrace}`,
      );
      break;
    }
    seen.set(num, j);
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main();
}
